### PR TITLE
Fix issues on CA1836 Fixer

### PR DESCRIPTION
**Pull request by Jozkee**
Follow-up to the recently added rule CA1836 on https://github.com/dotnet/roslyn-analyzers/pull/3584, the case `collection.Count().Equals(0)` or `collection.Count.Equals(0)` were causing an assertion failure on the Fixer. The reason is that the Fixer was not aware of the support for `collection.Count.Equals(0)` or `0.Equals(collection.Count)` invocation.

This PR addresses said scenarios on the fixer side. 

It also adds unit tests for `Concurrent`, `Immutable` and `Span` on CA1827, CA1829 and CA1836. 

EDIT:
Fix issues found when I ran the analyzer on [dotnet/runtime](https://github.com/dotnet/runtime)
* Fix error on Fixer when the expression was an argument.
* Fix issue on Fixer when the expression had an explicit cast.
* Avoid diagnosis when the operation is within the to-be-suggested `IsEmpty` property getter to prevent circularity.
* Fix trivia (missing whitespace and/or new line on the suggested fix).

### CODE REVIEW

File: src/NetAnalyzers/CSharp/Microsoft.NetCore.Analyzers/Performance/CSharpPreferIsEmptyOverCount.Fixer.cs [hunk 1]
```diff
@@ -15,26 +15,62 @@ namespace Microsoft.NetCore.CSharp.Analyzers.Performance
     [ExportCodeFixProvider(LanguageNames.CSharp), Shared]
     public sealed class CSharpPreferIsEmptyOverCountFixer : PreferIsEmptyOverCountFixer
     {
-        protected override SyntaxNode? GetExpressionFromMemberAccessor(SyntaxNode node)
+        protected override SyntaxNode? GetObjectExpressionFromOperation(SyntaxNode node, string operationKey)
         {
-            while (node is ParenthesizedExpressionSyntax parenthesizedExpression)
+            SyntaxNode? countNode = null;
+            switch (operationKey)
```
Comment (Evangelink) [inline src/NetAnalyzers/CSharp/Microsoft.NetCore.Analyzers/Performance/CSharpPreferIsEmptyOverCount.Fixer.cs:21]: What about using the new switch pattern here?

File: src/NetAnalyzers/UnitTests/Microsoft.NetCore.Analyzers/Performance/DoNotUseCountWhenAnyCanBeUsedTests.Base.cs [hunk 2]
```diff
@@ -65,5 +65,29 @@ protected async Task VerifyAsync(string methodName, string testSource, string fi
...
+            {
+                await this.Verifier.VerifyAsync(
+                    methodName,
+                    new string[] { testSource, extensionsSource },
+                    new string[] { fixedSource, extensionsSource },
+                    line, column);
+            }
+            catch
+            {
+                this.Output.WriteLine($"{SourceProvider.CommentPrefix} Source code:{Environment.NewLine}{testSource}");
```
Comment (Evangelink) [inline src/NetAnalyzers/UnitTests/Microsoft.NetCore.Analyzers/Performance/DoNotUseCountWhenAnyCanBeUsedTests.Base.cs:80]: Is it meant to be kept? This seems wrong to me.

File: src/NetAnalyzers/UnitTests/Microsoft.NetCore.Analyzers/Performance/DoNotUseCountWhenAnyCanBeUsedTests.Code.cs [hunk 3]
```diff
@@ -57,6 +57,10 @@ internal string WithDiagnostic(string code)
             internal string GetFixedExpressionCode(bool withPredicate, bool negate)
                 => $@"{GetLogicalNotText(negate)}{GetTargetExpressionCode(withPredicate, "Any" + this.MethodSuffix)}";
 
+
```
Comment (Evangelink) [inline src/NetAnalyzers/UnitTests/Microsoft.NetCore.Analyzers/Performance/DoNotUseCountWhenAnyCanBeUsedTests.Code.cs:60]: [cosmetic] Un-needed blank line.

**Comment by Jozkee** [comment]
The failing test on CI is marked with an `ActiveItem` attribute linking to https://github.com/dotnet/roslyn-analyzers/issues/3700 but that seems to not prevent execution.

On dotnet/runtime we use the `ActiveIssue` attribute and that helps to avoid these kind of CI errors, but `ActiveIssue` ships with a different package `Microsoft.DotNet.XUnitExtensions` that is not used by the unit test projects in this repo.

Is there precedence of how to handle this in roslyn-analyzers?

**Comment by mavasani** [comment]
I do not know about it. @jmarolf @dotnet/roslyn-analysis anyone?

**Comment by mavasani** [comment]
> The failing test on CI is marked with an WorkItem attribute linking to #3700 but that seems to not prevent execution.

You can mark the test with `[Fact(Skip = ...)]` to skip tests

**Comment by jmarolf** [comment]
we typically do `[Fact(Skip = "<url of github issue>"]` to point to the reason a test is being skipped

File: src/NetAnalyzers/CSharp/Microsoft.NetCore.Analyzers/Performance/CSharpPreferIsEmptyOverCount.Fixer.cs (hunk 1 above)
Comment (Jozkee) [reply src/NetAnalyzers/CSharp/Microsoft.NetCore.Analyzers/Performance/CSharpPreferIsEmptyOverCount.Fixer.cs:21]: We don't use one-liners or single calls here to determine `countNode`, I don't see it too beneficial. 

Or perhaps I don't see it the way you do, can you provide a snippet?

File: src/NetAnalyzers/UnitTests/Microsoft.NetCore.Analyzers/Performance/DoNotUseCountWhenAnyCanBeUsedTests.Base.cs (hunk 2 above)
Comment (Jozkee) [reply src/NetAnalyzers/UnitTests/Microsoft.NetCore.Analyzers/Performance/DoNotUseCountWhenAnyCanBeUsedTests.Base.cs:80]: Yeah, good call, it was probably there for debugging purposes during development. I will remove it and all code related.

cc @paulomorgado

**Comment by Jozkee** [comment]
Is there a way to generate the code coverage report locally?

**Comment by jmarolf** [comment]
> Is there a way to generate the code coverage report locally?
@sharwell

File: src/NetAnalyzers/CSharp/Microsoft.NetCore.Analyzers/Performance/CSharpPreferIsEmptyOverCount.Fixer.cs (hunk 1 above)
Comment (Evangelink) [reply src/NetAnalyzers/CSharp/Microsoft.NetCore.Analyzers/Performance/CSharpPreferIsEmptyOverCount.Fixer.cs:21]: I was thinking about something like

```csharp
var countNode = operationKey switch
{
	UseCountProperlyAnalyzer.OperationBinaryLeft when (node is BinaryExpressionSyntax binary) => binary.Left,
    UseCountProperlyAnalyzer.OperationBinaryRight when (node is BinaryExpressionSyntax binary) => 	binary.Right,
   UseCountProperlyAnalyzer.OperationEqualsArgument when (node is InvocationExpressionSyntax invocation) => invocationExpression.ArgumentList.Arguments[0].Expression,
	UseCountProperlyAnalyzer.OperationEqualsInstance when (node is InvocationExpressionSyntax invocation && invocation.Expression is MemberAcessExpressionSyntax memberAccess) => memberAccess.Expression,
	_ => // Not sure what you want to do here
};
```

File: src/NetAnalyzers/UnitTests/Microsoft.NetCore.Analyzers/Performance/DoNotUseCountWhenAnyCanBeUsedTests.Base.cs (hunk 2 above)
Comment (paulomorgado) [reply src/NetAnalyzers/UnitTests/Microsoft.NetCore.Analyzers/Performance/DoNotUseCountWhenAnyCanBeUsedTests.Base.cs]: @Jozkee,
If I was the one that did it, I don't recall it at the moment.

**Comment by mavasani** [comment]
@Jozkee I am seeing the following exception when attempting to FixAll `CA1836` violations on roslyn-analyzers repo. Is this something that will be addressed by this PR, or would you like me to file an issue?

```
System.InvalidCastException : Unable to cast object of type 'Microsoft.CodeAnalysis.CSharp.Syntax.ArgumentSyntax' to type 'Microsoft.CodeAnalysis.CSharp.Syntax.BinaryExpressionSyntax'.
   at Microsoft.NetCore.CSharp.Analyzers.Performance.CSharpPreferIsEmptyOverCountFixer.GetMemberAccessorFromBinary(SyntaxNode binaryExpression,Boolean useRightSide)
   at async Microsoft.NetCore.Analyzers.Performance.PreferIsEmptyOverCountFixer.<>c__DisplayClass3_0.<RegisterCodeFixesAsync>b__0(<Unknown Parameters>)
   at async Microsoft.CodeAnalysis.CodeActions.CodeAction.GetChangedSolutionAsync(<Unknown Parameters>)
   at async Microsoft.CodeAnalysis.CodeActions.CodeAction.GetChangedSolutionInternalAsync(<Unknown Parameters>)
   at async Microsoft.CodeAnalysis.CodeFixes.BatchFixAllProvider.GetChangedDocumentsAsync(<Unknown Parameters>)
   at async Microsoft.CodeAnalysis.CodeFixes.BatchFixAllProvider.GetDocumentIdToChangedDocumentsAsync(<Unknown Parameters>)
   at async Microsoft.CodeAnalysis.CodeFixes.BatchFixAllProvider.TryMergeFixesAsync(<Unknown Parameters>)
   at async Microsoft.CodeAnalysis.CodeFixes.BatchFixAllProvider.TryGetMergedFixAsync(<Unknown Parameters>)
   at async Microsoft.CodeAnalysis.CodeFixes.BatchFixAllProvider.GetFixAsync(<Unknown Parameters>)
   at async Microsoft.CodeAnalysis.CodeFixes.BatchFixAllProvider.GetFixAsync(<Unknown Parameters>)
   at async Microsoft.CodeAnalysis.Editor.Implementation.Suggestions.FixAllGetFixesService.GetFixAllCodeActionAsync(<Unknown Parameters>)
   at async Microsoft.CodeAnalysis.Editor.Implementation.Suggestions.FixAllGetFixesService.GetFixAllOperationsAsync(<Unknown Parameters>)
   at async Microsoft.CodeAnalysis.CodeActions.CodeAction.GetOperationsCoreAsync(<Unknown Parameters>)
   at Roslyn.Utilities.TaskExtensions.WaitAndGetResult_CanCallOnBackground[T](Task`1 task,CancellationToken cancellationToken)
   at Microsoft.CodeAnalysis.Editor.Implementation.Suggestions.SuggestedAction.InvokeWorker(Func`1 getFromDocument,IProgressTracker progressTracker,CancellationToken cancellationToken)
   at Microsoft.CodeAnalysis.Editor.Implementation.Suggestions.SuggestedAction.<>c__DisplayClass20_0.<InvokeCore>b__0()
   at Microsoft.CodeAnalysis.Extensions.IExtensionManagerExtensions.PerformAction(IExtensionManager extensionManager,Object extension,Action action)
   at Microsoft.VisualStudio.Telemetry.WindowsErrorReporting.WatsonReport.GetClrWatsonExceptionInfo(Exception exceptionObject)
```

**Comment by mavasani** [comment]
I have filed https://github.com/dotnet/roslyn-analyzers/issues/3798 as the exception is blocking enabling the rule as a warning on the repo due to 200+ violations, which would need FixAll.

**Comment by Jozkee** [comment]
@mavasani that should be fixed by this PR.
>Fix error on Fixer when the expression was an argument.

**Comment by Jozkee** [comment]
Merged with master and converged with changes made in https://github.com/dotnet/roslyn-analyzers/pull/3729.

**Comment by mavasani** [comment]
Thanks @Jozkee!

**Pull request merged**